lib/science_wire_harvester.rb

```diff
@@ -443,24 +189,271 @@ def harvest_from_wos_id_cap_profile_id_report(path_to_report)
+    # and `PubmedSourceRecord` into the `Publication.pub_hash`. The `SciencewireSourceRecord`
+    # data include a PubMed ID (`pmid`) so that we can link the two records.
+    def process_queued_pubmed_records
+      return if @records_queued_for_pubmed_retrieval.empty?
+      begin
+        pubmed_source_record = PubmedSourceRecord.new
+        pub_med_records = @pubmed_client.fetch_records_for_pmid_list(@records_queued_for_pubmed_retrieval.keys)
+        Nokogiri::XML(pub_med_records).xpath('//PubmedArticle').each do |pub_doc|
+          pmid = pub_doc.xpath('MedlineCitation/PMID').text
+          pubmed_source_record = PubmedSourceRecord.create_pubmed_source_record(pmid, pub_doc)
+          @total_new_pubmed_source_count += 1 if pubmed_source_record
+          pub_hash = @records_queued_for_pubmed_retrieval[pmid][:sw_hash]
+          author_ids = @records_queued_for_pubmed_retrieval[pmid][:authors]
+          pub = create_new_harvested_pub(pub_hash[:sw_id], pmid)
+          abstract = pubmed_source_record.extract_abstract_from_pubmed_record(pub_doc)
+          mesh = pubmed_source_record.extract_mesh_headings_from_pubmed_record(pub_doc)
+
+          pub_hash[:mesh_headings] = mesh unless mesh.blank?
+          pub_hash[:abstract] = abstract unless abstract.blank?
+
+          create_contribs_for_author_ids_and_pub(author_ids, pub)
+          pub.pub_hash = pub_hash
+          pub.sync_publication_hash_and_db
+          pub.save
+        end
+      rescue => e
+        NotificationManager.error(e, 'PubMed harvesting failed', self)
+      end
+      @records_queued_for_pubmed_retrieval.clear
+    end
+
+    def process_bibtex_file(path_to_file)
+      mode = nil
+      ids = []
+      IO.readlines(path_to_file).each do |l|
+        if l =~ /^@inproceedings/
+          mode = :inproceedings
+          next
+        end
+        next unless l =~ /^Unique-ID.*ISI:(.*)}},/
+        if mode == :inproceedings
+          @wos_ids_processed += 1
+          ids << Regexp.last_match(1)
+        end
+        mode = nil
+      end
+      ids
+    end
+
+    # @param [String] line A string with the format "WOS:000314860700023	38808"
+    # @return [Array<String>] first element wos_id. second_element cap_profile_id
+    def parse_wos_line(line)
+      line =~ /WOS:([^\s]*)\s+(\d+)/
+      [Regexp.last_match(1), Regexp.last_match(2)]
+    end
+
+    def process_wos_ids_for_author(cap_profile_id, wos_ids)
+      sunetid = Author.where(cap_profile_id: cap_profile_id).pluck(:sunetid).first
+      if sunetid.nil?
+        logger.error "No sunetid found for Author.find_by(cap_profile_id: #{cap_profile_id}). Skipping"
+      else
+        logger.info "Starting harvest for Author.find_by(sunetid: #{sunetid}, cap_profile_id: #{cap_profile_id}) with #{wos_ids.size} WOS ids"
+        harvest_sw_pubs_by_wos_array_and_sunetid sunetid, wos_ids
+      end
+    end
 
     def logger
       @logger ||= NotificationManager.sciencewire_logger
```

Comment:
Convert to attr_reader with init in the initializer?

Reply:
This is lazier (therefore better), since logger might not get called for a given set of operations.